### PR TITLE
Feature/sort index

### PR DESCRIPTION
**Pull request by jordanmaguire**
This is part of some functionality that I'll be using in El Mulcho.

Adding a sorting functionality to the users#index. This should be able to be used for any attribute on a model.

Off:

screen shot 2015-05-22 at 9 45 46 am

First click:

screen shot 2015-05-22 at 9 45 37 am

Second click:

screen shot 2015-05-22 at 9 45 43 am

### CODE REVIEW

File: lib/model_sorter.rb [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+      scope
+    end
+  end
+
+private
+
+  def self.sort_scope(scope, attribute, direction)
+    case direction
+    when "asc", "desc"
+      scope.order("#{attribute} #{direction}")
```
Comment (osiro) [inline lib/model_sorter.rb]: I believe there's a SQL injection vulnerability in here, check this out: http://rails-sqli.org/#order

A workaround would be sanitising the attribute by checking if that really belongs to one of those model attrs.

Comment (osiro) [inline lib/model_sorter.rb]: Yeap, just confirmed and it works, look:

```ruby
a = "(CASE SUBSTR(encrypted_password, 1, 1) WHEN 's' THEN 0 else 1 END)"
User.order("#{a} ASC")

# =>  SELECT "users".* FROM "users" ORDER BY (CASE SUBSTR(encrypted_password, 1, 1) WHEN 's' THEN 0 else 1 END) ASC

User.order(a => :asc)
# => ArgumentError: Input string is longer than NAMEDATALEN-1 (63)
```

Comment (jordanmaguire) [inline lib/model_sorter.rb]: Will fix - didn't realise you could do `order(attribute => direction)`. That is new in 4 by the looks of it. (didn't work in tempo which is 3)

Comment (osiro) [inline lib/model_sorter.rb]: Yeah I think it's new.

**Comment by osiro** [comment]
:+1:

File: app/controllers/admin/users_controller.rb [hunk 2]
```diff
@@ -2,12 +2,12 @@ class Admin::UsersController < Admin::BaseController
 
   def index
     authorize(User)
-    @users = policy_scope(User.member).page(params[:page])
+    @users = sort(policy_scope(User.member).page(params[:page]))
```
Comment (swebb) [inline app/controllers/admin/users_controller.rb]: I feel like I'm missing something. Don't you need to do the sort before you do the pagination?

Comment (jordanmaguire) [inline app/controllers/admin/users_controller.rb]: it's just a scope so it doesn't matter in which order you do it. I'll change it around so it reads better, though

Comment (osiro) [inline app/controllers/admin/users_controller.rb]: It actually doesn't matter, AR manages to place the limit and order in the right places.

```ruby
User.limit(10).order(id: :asc)
# => SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT 10

User.order(id: :asc).limit(10)
# => SELECT "users".* FROM "users" ORDER BY "users"."id" ASC LIMIT 10
```

File: config/application.rb [hunk 3]
```diff
@@ -8,6 +8,8 @@
 
 module CurtinVolunteers
```
Comment (pzi) [inline config/application.rb]: Should be Rails template?

Comment (jordanmaguire) [inline config/application.rb]: snap

**Comment by swebb** [comment]
:+1:

File: spec/helpers/application_helper_spec.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+
+  describe "#sortable_heading_for" do
+    subject { helper.sortable_heading_for(attribute) }
+    let(:attribute) { "dongle" }
+
+    before do
+      stub_request = Object.new
+      allow(stub_request).to receive(:path).and_return("x")
+      allow(stub_request).to receive(:query_parameters).and_return({sort_direction: "asc"})
+      allow(helper).to receive(:request).and_return(stub_request)
```
Comment (swebb) [inline spec/helpers/application_helper_spec.rb]: I think we should enable verification of partial doubles.

RSpec.configure do |config|
  config.mock_with :rspec do |mocks|
    mocks.verify_partial_doubles = true
  end
end

Comment (jordanmaguire) [inline spec/helpers/application_helper_spec.rb]: Done

**Pull request merged**